crates/sparrow-compiler/src/types/inference.rs

```diff
@@ -64,16 +67,71 @@ pub(crate) fn instantiate(
     // chain).
     let parameter_types = arguments.vec_for_varargs(parameters.types());
 
+    // Contains the fenl type(s) for each type variable.
+    let mut types_for_variable: HashMap<TypeVariable, Vec<Located<FenlType>>> = HashMap::new();
+
     for (index, (argument_type, parameter_type)) in
-        izip!(arguments.iter(), &parameter_types).enumerate()
+        izip!(arguments.iter(), parameter_types.clone()).enumerate()
     {
         match (parameter_type.inner(), argument_type.inner()) {
-            (FenlType::Generic(type_var), _) => {
+            (FenlType::TypeRef(type_var), _) => {
                 types_for_variable
                     .entry(type_var.clone())
                     .or_default()
-                    .push(argument_type);
+                    .push(argument_type.clone());
             }
+            (FenlType::Collection(_, type_vars), arg_type) => match arg_type {
```

Comment:
Here's where I am getting the `type` for each type variable within the `FenlType::Collection` (in this case, just the `map` is supported). 

The `arg_type` is parsed as a `DataType::Struct(..)`, which makes sense, since the parquet schema is a `record`:

```
{
    "name" : "map",
    "type" : [ "null", {
      "type" : "record",
      "name" : "map",
      "fields" : [ {
        "name" : "f1",
        "type" : [ "null", "string" ],
        "default" : null
      }, {
        "name" : "f2",
        "type" : [ "null", "string" ],
        "default" : null
      } ]
    } ],
    "default" : null
  },
```

Reply:
But then it's a bit weird, in that I don't have the type for the `key` which theoretically should be able to be any non-complex type. It's just assumed it's a `string` since it's in a `struct` (#120)

Reply:
I think this part probably needs to be moved to a helper function and/or struct. Basically, we need to be able to recursively handle both sides -- the signature ("parameters") and the actual types being used in the inference ("arguments").

The following assumes a function -- we could just as easily have a `struct` that wraps the `types_for_variable`, which may make it *easier* since we need to do things like the "add this fenl type to the vector".

I'll also say that there are likely *much* easier ways of handling location here, but it would be a pretty painful refactoring. If you hit too much pain with locations, let's discuss.

```
fn unify_types(
    parameter: &FenlType,
    argument: &FenlType,
    types_for_variable: &mut HashMap<TypeVariable, Vec<Located<FenlType>>>,
) -> error_stack::Result<(), Error> {
  match (parameter, argument) {
    (FenlType::TypeRef(variable), lhs) => {
      // Add `lhs` to the types_for_variable[variable]`
    }
    (_, FenlType::TypeRef(_)) => {
      // error -- type variables shouldn't appear in the argumetns
    }
    (FenlType::Collection(p_name, p_types), FenlType::Collection(a_name, a_types)) => {
      error_stack::ensure!(a_name == p_name, /* error -- can't use incompatible collections*);
      error_stack::ensure!(p_types.len() == a_types.len(), Error::Internal);
      // Recursively apply to each of the arguments.
      for (p, a) in p_types.zip(a_types) {
        unify_types(p, a, types_for_variable)?;
      }
     ...
     // other cases:
     // Equal concrete types vs. unequal concrete types.
     //
     // Parameter is a collection, right-hand-side is concrete.
     // Likely a bit painful, but I *think* we could actually have a "as collection" type thing.
     // Basically, we could treat `FenlType::Concrete(DataType::List(field))` as
     // `FenlType::Collection("list", vec![field.data_type()])` which would let us
     // handle it using the same code as above. Maybe we need a helper or something.
     // May be a reason to introduce the struct.
  }
}
```

Reply:
Let's look tomorrow -- I actually think it may have inferred the wrong type for the parquet columns as well.

Reply:
Does what exists now sufficiently handle the non-nested case though? 

The main question I have here is if the `arguments` type is correct and is inferred to be a `DataType::Struct`, then I don't have the `key` type, so unclear how we would support non-string types. I'll keep on trying to go from `batch -> parquet` and see if that column is interpreted differently.

Reply:
Current status -- current impl handles non-nested case, which I want to say is sufficient for now. I'd like to get started on the implementation for the primitive types, then can come back to support nested types and complex-type implementation afterwards.